### PR TITLE
Plan Label, Branching, and HTTP Header Support

### PR DESCRIPTION
**Pull request by twistedpair**
- Add support to trigger tests by plan label.
- Add support to trigger tests on mabl branch
- Add support to add http-headers to tests
- Improved input validation
- Refactor/simplify, more test coverage

Will go out as `v1.10`

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -32,6 +32,50 @@ jobs:
           environment-id: <your-environment-id-e>
 ```
 
+### Example workflow: Complete
```
Comment (twistedpair) [inline README.md:35]: Added "complete" example.

File: src/index.ts [hunk 2]
```diff
@@ -23,45 +25,65 @@ const EXECUTION_COMPLETED_STATUSES = [
 ];
 const GITHUB_BASE_URL = 'https://api.github.com';
 
-async function run(): Promise<void> {
+export function optionalArrayInput(name: string): string[] {
```
Comment (twistedpair) [inline src/index.ts:28]: Helpers, tested.

File: src/index.ts [hunk 3]
```diff
@@ -23,45 +25,65 @@ const EXECUTION_COMPLETED_STATUSES = [
...
-      required: false,
-    });
-    const environmentId: string = core.getInput('environment-id', {
-      required: false,
-    });
+    const applicationId = optionalInput(ActionInputs.ApplicationId);
+    const environmentId = optionalInput(ActionInputs.EnvironmentId);
 
-    const apiKey: string = process.env.MABL_API_KEY || '';
+    const apiKey = process.env.MABL_API_KEY;
```
Comment (twistedpair) [inline src/index.ts:66]: We should bail if no key is defined, rather than wait for API call to auth error.

File: src/index.ts [hunk 4]
```diff
@@ -101,28 +127,40 @@ async function run(): Promise<void> {
...
+      uri,
+      revision,
+      mablBranch,
     );
 
-    core.setOutput('mabl-deployment-id', deployment.id);
+    core.setOutput(ActionOutputs.DeploymentId, deployment.id);
 
-    let outputLink: string = baseApiUrl;
+    let appOrEnv: Application | Environment | undefined;
```
Comment (twistedpair) [inline src/index.ts:146]: Fail out if neither app-id or env-id are valid and supplied.

File: src/mablApiClient.ts [hunk 5]
```diff
@@ -83,27 +96,33 @@ export class MablApiClient {
...
-    uri: string,
+    browserTypes: string[],
+    planLabels: string[],
+    httpHeaders: string[],
     rebaselineImages: boolean,
     setStaticBaseline: boolean,
-    revision: string | undefined,
     eventTime: number,
     properties: DeploymentProperties,
+    applicationId?: string,
```
Comment (twistedpair) [inline src/mablApiClient.ts:106]: Move optional params to end of signature.

File: src/index.ts (hunk 4 above)
Comment (simonychoy) [reply src/index.ts:146]: It looks like as long as the application ID is valid, the environment ID does not get checked. Does the Deployment entity contain info about the number of plans/tests that are going to be started?

**Review by simonychoy — APPROVED** [review]
Very neat!

File: src/index.ts (hunk 4 above)
Comment (twistedpair) [reply src/index.ts:146]: Thanks, @simonychoy . This validation pass is confirming that we have at least one valid entity, to build the UI app URL. The backing REST API will conduct a more thorough check.

**Pull request merged**